src/xdg-autostart-generator/xdg-autostart-generator.c

```diff
@@ -0,0 +1,120 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+
+#include <errno.h>
+#include <stdio.h>
+#include <unistd.h>
+
+#include "xdg-autostart-service.h"
+#include "dirent-util.h"
+#include "fd-util.h"
+#include "generator.h"
+#include "hashmap.h"
+#include "log.h"
+#include "main-func.h"
+#include "path-lookup.h"
+#include "stat-util.h"
+#include "string-util.h"
+#include "nulstr-util.h"
+#include "strv.h"
+
+static void free_xdgautostartservice_hashmapp(Hashmap **h) {
+        hashmap_free_with_destructor(*h, xdg_autostart_service_free);
+}
+
+static int enumerate_xdg_autostart(Hashmap *all_services) {
+        _cleanup_strv_free_ char **autostart_dirs = NULL;
+        _cleanup_strv_free_ char **config_dirs = NULL;
+        _unused_ _cleanup_strv_free_ char **data_dirs = NULL;
+        _cleanup_free_ char *user_config_autostart_dir = NULL;
+        char **path;
+        int r;
+
+        r = xdg_user_config_dir (&user_config_autostart_dir, "/autostart");
+        if (r < 0)
+                return r;
+        r = strv_extend (&autostart_dirs, user_config_autostart_dir);
+        if (r < 0)
+                return r;
+
+
+        r = xdg_user_dirs (&config_dirs, &data_dirs);
+        if (r < 0)
+                return r;
+        r = strv_extend_strv_concat (&autostart_dirs, config_dirs, "/autostart");
+        if (r < 0)
+                return r;
+
+
+        STRV_FOREACH(path, autostart_dirs) {
+                _cleanup_closedir_ DIR *d = NULL;
+                struct dirent *de;
+
+                d = opendir(*path);
+                if (!d) {
+                        if (errno != ENOENT)
+                                log_warning_errno(errno, "Opening %s failed, ignoring: %m", *path);
+                        continue;
+                }
+
+                FOREACH_DIRENT(de, d, log_error_errno(errno, "Failed to enumerate directory %s, ignoring: %m", *path)) {
```

Comment:
log_warning_errno (since it's not fatal).

Reply:
This is a direct copy from the sysv generator. Maybe it should also be changed there?

```
                        FOREACH_DIRENT(de, d, log_error_errno(errno, "Failed to enumerate directory %s, ignoring: %m", path)) {
```

Reply:
+1